File: 0020-valid-parentheses/0020-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    let stack = [];
    for (let i=0; i<s.length; i++) {
        switch (s.charAt(i)) {
            case "(":
                stack.push(0);
                break;

            case "{":
                stack.push(1);
                break;

            case "[":
                stack.push(2);
                break;

            case ")":
                if (stack.at(-1) === 0) stack.pop();
                else return false;
                break;
            
            case "}":
                if (stack.at(-1) === 1) stack.pop();
                else return false;
                break;

            case "]":
                if (stack.at(-1) === 2) stack.pop();
                else return false;
                break;

            default:
                return false;
        }
    }
    return stack.length === 0
};

File: 0020-valid-parentheses/0020-valid-parentheses.kt
class Solution {
    fun isValid(s: String): Boolean {
        val stack = ArrayDeque<Char>()
        val parenthesisMap = mapOf<Char, Char> (
            ')' to '(',
            ']' to '[',
            '}' to '{'
        )
        for (c in s) {
            if (c in parenthesisMap) {
                val poppedValue = stack.removeLastOrNull()
                if (poppedValue != parenthesisMap.getValue(c)) {
                    println(poppedValue)
                    return false
                }
                continue
            }
            stack.addLast(c)
        }
        return stack.isEmpty()
    }
}